### DISCUSSION TITLE
Field meta data, revisited

### DISCUSSION THREAD

**Discussion by msageryd** [react-hook-form/react-hook-form#7559] [Q&A]
I found this discussion on this topic, but no conclusion.
https://github.com/react-hook-form/react-hook-form/discussions/3975

I need to propagate some metadata for each field down to each field component. Is there any way to attach metadata to a field?

As of now I have a separate object, keyed with each fieldname, which holds metadata for each field. This object is passed to the fields via context. It would be cleaner to attach metadata to each field from start.

I get the field metadata from my API.

This is how it could look like:
```
register(field.fieldName, {meta: field.meta, etc})
```

**Comment by msageryd** [comment]
Maybe I should clarify my use case.

My forms are completely driven from a JSON definition. The JSON includes:

**Field definitions**
- datatypes
- warning texts  (or rather i18next keys)
- error texts
- help texts
- field labels
- field icon names
- field validations

**Form layout**
- sort order
- component layout
- visual component types
- container definitions for "next levels", i.e. sub-forms

**Conditions**
- cross field rules
- ex: `when field A == true then fieldB.iconName = exclamation`
- ex: `when Field A == true or Field B == "abc" then fieldC.visible=false`

**So..**
- Based on the fieldDefinitions I register all fields via `register` when I start the form.
- Metadata for each field goes into a metadata-object, but I'd rather attch it to each field
- The form layout defines when to mount actual input components for the registered fields
- Conditions goes into a "conditions engine" which checks conditions based on useWatch() for all related fields
- When a condition rule gets activated, the result is put into a conditionObject which gets propagated via context.

There are three parts here:
1. static field metadata
2. conditions/rules
3. result from conditions

Part 1 would suit perfectly for attaching via the register function.

**Reply by Noitidart** [reply]
I came up with this solution, it works, but very implementation detail dependent - https://github.com/react-hook-form/react-hook-form/discussions/8161#discussion-3999584
